[NNC] enable fusion of conv with elementwise OP

chunyuan-w opened the pull request:
## Pitch
Enable Conv-Eltwise fusion in NNC.

## Description
This PR adds a `FuseConvWithEltwise` pass to fuse convolution with elementwise OP for TE subgraph. This pass will insert prepack and packed run ops for conv2d and enable fusion of conv2d with elementwise OPs. The fused packed run ops is implemented via external call in NNC.

## Code structure
Graph rewrite pass related code is placed in:
```
torch/csrc/jit/passes/mkldnn_rewrite.h
torch/csrc/jit/passes/mkldnn_rewrite.cpp
```

NNC integration of fused conv-eltwise OP via external call is located in:
```
torch/csrc/jit/tensorexpr/kernel.cpp

torch/csrc/jit/tensorexpr/operators/conv2d.h
torch/csrc/jit/tensorexpr/operators/conv2d.cpp

torch/csrc/jit/tensorexpr/lowerings.cpp
torch/csrc/jit/tensorexpr/external_functions.cpp
```

Fused prepack OP context is in:
```
aten/src/ATen/native/mkldnn/Common.h
aten/src/ATen/native/mkldnn/RegisterMkldnnOpContextClass.cpp
aten/src/ATen/native/mkldnn/OpContext.h
aten/src/ATen/native/mkldnn/OpContext.cpp
```

Fused OP implementation is done in:
```
aten/src/ATen/native/mkldnn/ConvPrepack.h
aten/src/ATen/native/mkldnn/ConvPrepack.cpp
```

## OP benchmark for conv-relu
The below performance is measured on top of these two PRs to support NHWC: https://github.com/pytorch/pytorch/pull/76948 and https://github.com/pytorch/pytorch/pull/78238.

- Measured on Cascade Lake 8280
- Jemalloc enabled
- batch_size = 1
- Channels Last format

### Single thread:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/chunyuan/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/chunyuan/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">



shape | time (us)_no_fusion | time (us)_fusion | Gain
-- | -- | -- | --
kernel=3, N=1, iC=64, H=56, W=56, oC=64,   stride=1, pad=1, dilates=1, g=1 | 1706.22 | 1371.97 | 19.59%
kernel=1, N=1, iC=256, H=56, W=56,   oC=512, stride=2, pad=0, dilates=1, g=1 | 2499.28 | 1571.52 | 37.12%
kernel=3, N=1, iC=256, H=56, W=56,   oC=256, stride=1, pad=1, dilates=1, g=32 | 4169.52 | 2738.53 | 34.32%
kernel=3, N=1, iC=512, H=56, W=56,   oC=512, stride=2, pad=1, dilates=1, g=32 | 3998.77 | 3085.85 | 22.83%
kernel=1, N=1, iC=64, H=56, W=56, oC=64,   stride=1, pad=0, dilates=1, g=1 | 673.73 | 430.81 | 36.06%
kernel=1, N=1, iC=256, H=56, W=56, oC=64,   stride=1, pad=0, dilates=1, g=1 | 1101.87 | 801.07 | 27.30%
kernel=1, N=1, iC=256, H=56, W=56,   oC=256, stride=1, pad=0, dilates=1, g=1 | 4692.91 | 3116.13 | 33.60%
kernel=1, N=1, iC=512, H=28, W=28,   oC=512, stride=1, pad=0, dilates=1, g=1 | 3310.64 | 2503.39 | 24.38%



</body>

</html>


### 4 threads:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/chunyuan/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/chunyuan/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">



shape | time (us)_no_fusion | time (us)_fusion | Gain
-- | -- | -- | --
kernel=3, N=1, iC=64, H=56, W=56, oC=64,   stride=1, pad=1, dilates=1, g=1 | 360.07 | 321.21 | 10.79%
kernel=1, N=1, iC=256, H=56, W=56,   oC=512, stride=2, pad=0, dilates=1, g=1 | 391.49 | 323.17 | 17.45%
kernel=3, N=1, iC=256, H=56, W=56,   oC=256, stride=1, pad=1, dilates=1, g=32 | 536.4 | 465.97 | 13.13%
kernel=3, N=1, iC=512, H=56, W=56,   oC=512, stride=2, pad=1, dilates=1, g=32 | 674.98 | 616.32 | 8.69%
kernel=1, N=1, iC=64, H=56, W=56, oC=64,   stride=1, pad=0, dilates=1, g=1 | 160.97 | 70.05 | 56.48%
kernel=1, N=1, iC=256, H=56, W=56, oC=64,   stride=1, pad=0, dilates=1, g=1 | 215.81 | 182.6 | 15.39%
kernel=1, N=1, iC=256, H=56, W=56,   oC=256, stride=1, pad=0, dilates=1, g=1 | 658.45 | 576.97 | 12.37%
kernel=1, N=1, iC=512, H=28, W=28,   oC=512, stride=1, pad=0, dilates=1, g=1 | 702.18 | 566.39 | 19.34%



</body>

</html>



### 1 socket (28 cores):
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/chunyuan/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/chunyuan/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">



shape | time (us)_no_fusion | time (us)_fusion | Gain
-- | -- | -- | --
kernel=3, N=1, iC=64, H=56, W=56, oC=64,   stride=1, pad=1, dilates=1, g=1 | 149.92 | 103.78 | 30.78%
kernel=1, N=1, iC=256, H=56, W=56,   oC=512, stride=2, pad=0, dilates=1, g=1 | 192.76 | 110.87 | 42.48%
kernel=3, N=1, iC=256, H=56, W=56,   oC=256, stride=1, pad=1, dilates=1, g=32 | 160.67 | 127.24 | 20.81%
kernel=3, N=1, iC=512, H=56, W=56,   oC=512, stride=2, pad=1, dilates=1, g=32 | 212.45 | 180.55 | 15.02%
kernel=1, N=1, iC=64, H=56, W=56, oC=64,   stride=1, pad=0, dilates=1, g=1 | 114.57 | 50.58 | 55.85%
kernel=1, N=1, iC=256, H=56, W=56, oC=64,   stride=1, pad=0, dilates=1, g=1 | 198.64 | 70.6 | 64.46%
kernel=1, N=1, iC=256, H=56, W=56,   oC=256, stride=1, pad=0, dilates=1, g=1 | 281.35 | 155.8 | 44.62%
kernel=1, N=1, iC=512, H=28, W=28,   oC=512, stride=1, pad=0, dilates=1, g=1 | 262.15 | 162.94 | 37.84%



</body>

</html>


## UT
```
test/test_mkldnn_fusion.py
```

chunyuan-w committed:
add graph rewrite pass for conv eltwise fusion

chunyuan-w committed:
add OP context for conv eltwise fusion

chunyuan-w committed:
add impl of conv eltwise fusion

chunyuan-w committed:
add FuseConvWithEltwise into NNC

chunyuan-w committed:
add UT for conv eltwise fusion

chunyuan-w committed:
do not rewrite depthwise conv2d with mkldnn prepacked conv

chunyuan-w committed:
improve UT check

chunyuan-w committed:
set canFuseOnCPU to be true for UT

chunyuan-w committed:
fix lint

chunyuan-w committed:
support bias=False for aten::conv2d

chunyuan-w committed:
set plain_format to be false and reorder blocked dst back to public

chunyuan-w committed:
Merge branch 'master' into chunyuan/conv_eltwise_fusion

chunyuan-w committed:
remove unused variable

EikanWang commented:
I would appreciate it if you could comment on this PR. Because it is the infrastructure, this PR impacts how we implement the consecutive implementations.

ZolotukhinM commented on torch/csrc/jit/passes/mkldnn_rewrite.cpp:
```diff
@@ -0,0 +1,276 @@
+  FuseReluWithPackedOps(graph);
+  GRAPH_DEBUG(
+      "After FuseReluWithPackedOps, before FoldPrePackingOps\n", *graph);
+  FoldPrePackingOps(graph);
```
Can we use the standard ConstantPropagation pass here? Does it not work for some reason?

chunyuan-w replied:
Inside the ConstantPropagation pass, when calling `runNodeIfInputsAreConstant`, the 3rd parameter `db` is not provided thus the default value `nullptr` will be used:
https://github.com/pytorch/pytorch/blob/9e806619ccbd211b2501fd380ec7d1bd398ec67c/torch/csrc/jit/passes/constant_propagation.cpp#L180-L183
Default value of `db` is `nullptr`:
https://github.com/pytorch/pytorch/blob/9e806619ccbd211b2501fd380ec7d1bd398ec67c/torch/csrc/jit/passes/constant_propagation.h#L26-L29

However for Custom class, if `!db`, the constant propagation won't be functional:
https://github.com/pytorch/pytorch/blob/9e806619ccbd211b2501fd380ec7d1bd398ec67c/torch/csrc/jit/passes/constant_propagation.cpp#L107-L113

In our case, the prepacked OP that we're trying to do the constant propagation is a custom class thus it does not work.

Is this by design or an issue that need to be fixed in the standard ConstantPropagation pass? If it is an issue, I can submit another PR to fix it.

ZolotukhinM replied:
I think we should just pass `aliasDb_` to `runNodeIfInputsAreConstant`. @eellison, do you know why we're not doing it now?

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -1046,7 +1046,8 @@ class TensorExprFuser {
 
     if (node->kind() == aten::conv2d) {
-      if (!tensorexpr::conv2dIsSupportedJit(node)) {
+      if (!tensorexpr::conv2dIsSupportedJit(node) &&
+          !tensorexpr::mkldnnPrepackedConvIsSupportedJit(node)) {
```
Should the condition be `||` here?

chunyuan-w replied:
Here we want to let NNC selects `conv2d` into the fusion group if either `conv2dIsSupportedJit` (depthwise, 3x3 kernel) or `mkldnnPrepackedConvIsSupportedJit` returns true.

`conv2dIsSupportedJit` only returns true for depthwise conv2d of 3x3 kernel. For other shapes that mkldnn prepacked conv2d can provide performance boost, we still want these `conv2d` OPs to be rewritten.

ZolotukhinM commented on torch/csrc/jit/passes/mkldnn_rewrite.cpp:
```diff
@@ -0,0 +1,276 @@
+  return tt->sizes();
+}
+
+bool isContiguous(Value* v) {
+  auto const& tt = v->type()->cast<TensorType>();
+  if (!tt) {
+    return false;
+  }
+  if (!tt->isComplete()) {
+    return false;
+  }
+  auto const& sizes = tt->sizes().concrete_sizes();
+  auto const& strides = tt->strides().concrete_sizes();
+  if (!sizes || !strides) {
+    return false;
+  }
+  int ndims = (*sizes).size();
+  // TODO: only support 4D input for now
+  if (ndims != 4) {
+    return false;
+  }
+
+  // TODO: currently only support contiguous input. Channels last contiguous
+  // will be supported as a next step
+  return *strides == TensorType::contiguousStridesOf(*sizes);
+}
```
This function already exists here, let's reuse it (please feel free to move it to another file if that helps):
https://github.com/pytorch/pytorch/blob/c7e8de6e86c32b727a6dc05bee645896ccc5262e/torch/csrc/jit/tensorexpr/kernel.cpp#L211-L235

chunyuan-w replied:
Modified the code to reuse this function: https://github.com/pytorch/pytorch/pull/77157/commits/0fc328b519648bd91ef50b274810cd7da3e6e799

ZolotukhinM commented on torch/csrc/jit/passes/mkldnn_rewrite.cpp:
```diff
@@ -0,0 +1,276 @@
+  n->output()->replaceAllUsesWith(prepack_conv->output());
+}
+
+bool canFuseOnDevice(Value* v) {
```
Similarly, please try using existing functions from TE fuser when possible.

chunyuan-w replied:
Below are some differences compared with the existing `canFuseOnDevice` function:
https://github.com/pytorch/pytorch/blob/192aa3ad5f770214bd4ad26297deff08c88b239f/torch/csrc/jit/passes/tensorexpr_fuser.cpp#L845-L867

1. Currently mkldnn prepacked conv2d only supports CPU
2. mkldnn prepacked conv2d requires MKLDNN to be enabled

Would it be okay that we add `canFuseOnDevice` in `mkldnn_rewrite.cpp` due to the above differences?

ZolotukhinM replied:
So, all we do here now is checking that for tensor values the device is cpu and mkldnn is enabled. It is confusing to reuse the same name as the existing function `isFusableOnDevice` + `canFuseOnDevice` as these functions are intended to be used in a generic fuser that handles multiple devices + multiple fusible ops. I suggest renaming this function to something like "isTensorTypeCPU" and replace several functions with one. For the second part of the check - whether mkldnn is enabled - I suggest moving this check to the beginning of the pass and not check it on every node.

chunyuan-w replied:
Changed the function according to the comment.

ZolotukhinM commented on torch/csrc/jit/passes/mkldnn_rewrite.cpp:
```diff
@@ -0,0 +1,276 @@
+  insertPrePackedConvOp(graph->block());
+}
+
+void insertMkldnnPrePackedOps(std::shared_ptr<Graph>& graph) {
+  insertMkldnnPrePackedConv2dOp(graph);
+}
```
Do we expect more rewrites here? If we don't plan to add them in near future, I suggest calling `insertMkldnnPrePackedConv2dOp` directly without an extra wrapper above it.

chunyuan-w replied:
We plan to add `insertMkldnnPrePackedLinearOp` as next step.

ZolotukhinM commented on torch/csrc/jit/passes/mkldnn_rewrite.cpp:
```diff
@@ -0,0 +1,276 @@
+#else
+
+void FuseConvWithEltwise(std::shared_ptr<Graph>& graph) {
+  GRAPH_DUMP("MKLDNN Not enabled", graph);
```
There is no need to dump the entire graph if we are not doing anything with it. Logging the message seems useful, so maybe change it to `GRAPH_DEBUG` so that we don't need to print the graph along with it?

chunyuan-w replied:
Thanks for the suggestion. Changed to `GRAPH_DEBUG` here.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.h:
```diff
@@ -24,6 +24,8 @@ struct SmallSizeTPairHash {
 // Returns true if the TE fuser supports this conv2d.
 bool conv2dIsSupportedJit(const Node* node);
+// Returns true if the TE fuser supports this conv2d with mkldnn prepacked conv.
+bool mkldnnPrepackedConvIsSupportedJit(const torch::jit::Node* node);
```
Nit: let's drop `torch::jit::` for consistency with the code around.

chunyuan-w replied:
Removed `torch::jit::` to align with the code around.

ZolotukhinM commented on torch/csrc/jit/passes/mkldnn_rewrite.cpp:
```diff
@@ -0,0 +1,276 @@
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/ir/subgraph_matcher.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/constant_pooling.h>
+#include <torch/csrc/jit/passes/constant_propagation.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+#include <torch/csrc/jit/passes/fold_conv_bn.h>
+#include <torch/csrc/jit/passes/freeze_module.h>
+#include <torch/csrc/jit/passes/graph_rewrite_helper.h>
+#include <torch/csrc/jit/passes/inliner.h>
+#include <torch/csrc/jit/passes/mkldnn_rewrite.h>
+#include <torch/csrc/jit/passes/remove_dropout.h>
+#include <torch/csrc/jit/runtime/graph_executor_impl.h>
```
Do we need all these includes? Please remove those that are not actually used (e.g. `freeze_module.h`).

chunyuan-w replied:
Thanks for pointing that out. I've removed headers that are not actually needed in this file (https://github.com/pytorch/pytorch/pull/77157/commits/ebeb5792167eecbee051955080b4ab24c64a45a5)

ZolotukhinM commented on test/test_mkldnn_fusion.py:
```diff
@@ -0,0 +1,110 @@
+                x = self.eltwise(x)
+                return x
+
+        for eltwise in ["relu"]:
```
Would it be simpler if we had a list of functions themselves instead of a name by which we need to look them up?
I.e.:
```
for eltwise_fn in [torch.nn.functional.relu, ...]:
  ...
```

chunyuan-w replied:
Thanks for the suggestion! I've modified it to directly iterate through functions instead of looking them up (https://github.com/pytorch/pytorch/pull/77157/commits/aeaf507cf964559ad0003a5ea5ed0a6c44b2d50c).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/operators/conv2d.cpp:
```diff
@@ -1,3 +1,5 @@
+#include <ATen/Parallel.h>
```
Why do we need `ATen/Parallel.h`?

chunyuan-w replied:
Thanks for the reminder. I'll remove it.

chunyuan-w replied:
Have removed `ATen/Parallel.h`.

ZolotukhinM reviewed (commented):
Overall looks good, I have some comments below. Also, do you have any performance measurements for this? It would be good to include into the commit message

chunyuan-w committed:
use GRAPH_DEBUG when no change happens on the graph

chunyuan-w committed:
unify namespace usage with other code

chunyuan-w committed:
remove unused header file

chunyuan-w committed:
UT: directly iterate through eltwise func name

chunyuan-w committed:
Merge branch 'master' into chunyuan/conv_eltwise_fusion

chunyuan-w committed:
reuse isContiguous in existing file

chunyuan-w committed:
graph rewrite: check NHWC

chunyuan-w committed:
conv kernel: support nhwc

chunyuan-w committed:
add NHWC in UT

chunyuan-w committed:
remove ATen/Parallel.h and thread check

chunyuan-w committed:
refine GRAPH_DEBUG message

chunyuan-w committed:
UT: add more shapes

chunyuan-w committed:
remove unused ConvUtils.h

chunyuan-w committed:
add ASSERT in kernel: input should be ChannelsLast

chunyuan-w commented:
> Overall looks good, I have some comments below. Also, do you have any performance measurements for this? It would be good to include into the commit message

Thanks for the suggestion. I've updated the performance number in the PR description.

The performance is measured on top of https://github.com/pytorch/pytorch/pull/76948 and https://github.com/pytorch/pytorch/pull/78238 in order to support NHWC. If these two are landed, I'll then push new commits to support fusion with NHWC format in this PR. I'm not pushing those commits now since, without those two PRs, there will be building issues.

chunyuan-w committed:
change isFusableOnDevice to isTensorTypeCPU

chunyuan-w committed:
Merge branch 'master' into chunyuan/conv_eltwise_fusion

chunyuan-w committed:
Merge branch 'master' into chunyuan/conv_eltwise_fusion

chunyuan-w committed:
Merge branch 'master' into chunyuan/conv_eltwise_fusion

chunyuan-w commented:
https://github.com/pytorch/pytorch/pull/76948 and https://github.com/pytorch/pytorch/pull/78238 are merged and I've merged the latest change from the master branch.

chunyuan-w committed:
conv kernel: support nchw

chunyuan-w committed:
ut: add oC=1

chunyuan-w committed:
use compute_v2

chunyuan-w committed:
add outputStrides when creating ResultBuf

chunyuan-w commented:
It seems that ideep dependency change has been reverted in the master branch:
https://github.com/pytorch/pytorch/pull/78238 (landed and reverted) and https://github.com/pytorch/pytorch/pull/79258 (landed take-2 and reverted in https://github.com/pytorch/pytorch/commit/05624bcf7b831ed18b857701757da2a4e076c493).

chunyuan-w committed:
Merge branch 'master' into chunyuan/conv_eltwise_fusion

chunyuan-w committed:
Merge branch 'master' into chunyuan/conv_eltwise_fusion

ZolotukhinM reviewed (approved):
LGTM!

chunyuan-w committed:
fix BZL build

chunyuan-w committed:
Merge branch 'master' into chunyuan/conv_eltwise_fusion

chunyuan-w committed:
Merge branch 'master' into chunyuan/conv_eltwise_fusion

xsacha commented:
Needs a rebase after latest changes on master. Thanks!

chunyuan-w committed:
separate fusion_attr_map and fusion_rewrite_map

chunyuan-w committed:
Merge branch 'master' into chunyuan/conv_eltwise_fusion

ZolotukhinM commented:
@pytorchbot rebase -s

ZolotukhinM commented:
`ideep` seems to have been updated internally, so I'm going to try landing once more 🤞 sorry that it takes so long!

ZolotukhinM commented:
@chunyuan-w could you please rebase the PR on a recent master? Looks like there are some conflicts

chunyuan-w committed:
Merge branch 'master' into chunyuan/conv_eltwise_fusion

chunyuan-w commented:
@ZolotukhinM I've merged the recent master.
